File: Chapter 05/Python/section_5_6_crypto_tracker/after/core/domain/portfolio_manager.py
from core.ports.price_provider_port import PriceProviderPort

class PortfolioManager:
    """
    CORE – Pure business logic.
    No HTTP clients, no JSON parsing. This class is fully isolated.
    """

    def __init__(self, price_provider: PriceProviderPort):
        """Dependency Injection via Constructor."""
        self.price_provider = price_provider

    def calculate_total_value(self, btc_amount: float) -> float:
        # We just call the port. We don't care WHERE the price comes from.
        current_price = self.price_provider.get_bitcoin_price()
        return btc_amount * current_price

File: Chapter 05/JS/section_5_6_crypto_tracker/after/core/domain/portfolioManager.js
/**
 * CORE – Pure business logic.
 * No fetch calls, no JSON parsing. This class is fully isolated.
 */
class PortfolioManager {
    /**
     * Dependency Injection via Constructor.
     * @param {Object} priceProvider - An adapter that implements getBitcoinPrice()
     */
    constructor(priceProvider) {
        this.priceProvider = priceProvider;
    }

    async calculateTotalValue(btcAmount) {
        // We just call the port. We don't care WHERE the price comes from.
        const currentPrice = await this.priceProvider.getBitcoinPrice();
        return btcAmount * currentPrice;
    }
}

module.exports = PortfolioManager;